### PR TITLE
drivetrain & drive

### PR DESCRIPTION
**Pull request by Yoel-Perman-Brilliant**

### CODE REVIEW

File: src/main/java/frc/robot/commands/DriveSwerve.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+        if (Math.abs(xSpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) xSpeed = 0;
+        if (Math.abs(ySpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) ySpeed = 0;
+        if (Math.abs(rotationSpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) rotationSpeed = 0;
+
+        drivetrain.drive(xSpeed, ySpeed, rotationSpeed, fieldRelative, usePID);
+    }
+
+    @Override
+    public boolean isFinished() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/DriveSwerve.java]: remove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+            new Translation2d(TRACK_WIDTH / 2, -TRACK_WIDTH / 2);
+    public static final Translation2d BACK_LEFT_WHEEL_POSITION =
+            new Translation2d(-TRACK_WIDTH / 2, TRACK_WIDTH / 2);
+    public static final Translation2d BACK_RIGHT_WHEEL_POSITION =
+            new Translation2d(-TRACK_WIDTH / 2, -TRACK_WIDTH / 2);
+    public static final Translation2d CENTER_OF_ROBOT = new Translation2d(
+            (FRONT_LEFT_WHEEL_POSITION.getX() + BACK_RIGHT_WHEEL_POSITION.getX()) / 2,
+            (FRONT_LEFT_WHEEL_POSITION.getY() + BACK_RIGHT_WHEEL_POSITION.getY()) / 2);
+
+    public static final double MAX_SPEED_METERS_PER_SECONDS = 4;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: should check

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    public static final double MIN_SPEED_METERS_PER_SECONDS = 0.2;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:32]: ditto

File: src/main/java/frc/robot/subsystems/Drivetrain.java
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:1]: Everything here about pose estimation is very messy, I HIGHLY recommend seeing trigon's code and adapting it. It's a little complicated at first, but it's a lot more clean and easily allows for multiple cameras

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    /**
+     * Resets the position to the position obtained from the vision source, or the given position if there is no
+     * {@link AprilTag} target.
+     *
+     * @param noTargetPosition the position to reset the estimator to if there is no information from the vision source.
+     * @TODO add to {@link frc.robot.Robot}
+     */
+    public void resetPoseEstimator(Pose2d noTargetPosition) {
+//        poseEstimator.resetPosition(
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: also

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        setSpeed(state.speedMetersPerSecond, usePID);
+    }
+
+    private void configureDriveController() {
+        driveController.getPIDController().setP(drivePIDSettings.getkP());
+        driveController.getPIDController().setI(drivePIDSettings.getkI());
+        driveController.getPIDController().setD(drivePIDSettings.getkD());
+        driveController.setInverted(driveInverted);
+        driveController.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        driveEncoder.setPositionConversionFactor((1 / DRIVING_GEAR_RATIO) * WHEEL_CIRCUMFERENCE_METERS);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: move this to `configureDriveEncoder`

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+        driveEncoder.setVelocityConversionFactor(
+                ((1 / DRIVING_GEAR_RATIO) * WHEEL_CIRCUMFERENCE_METERS) / 60);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: image

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    private void configureAbsoluteEncoder() {
+        MagnetSensorConfigs config = new MagnetSensorConfigs();
+        config.withAbsoluteSensorRange(AbsoluteSensorRangeValue.Unsigned_0To1);
+        config.withSensorDirection(cancoderInverted ? SensorDirectionValue.Clockwise_Positive : SensorDirectionValue.CounterClockwise_Positive);
+    }
+
+    public void configureRelativeTurnEncoder() {
+        turnEncoder.setPositionConversionFactor((1 / STEERING_GEAR_RATIO) * 360);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: image

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+    public void setAngle(double angle) {
+        configureTurnController();
+        turnController.getPIDController().setReference(angle, CANSparkMax.ControlType.kPosition, PID_SLOT);
+    }
+
+    //speed - m/s
+    private void setSpeed(double speed, boolean usePID) {
+        if (usePID) {
+            configureDriveController();
+            driveFeedForwardController.setGains(driveFeedForwardSettings.getkS(), driveFeedForwardSettings.getkV(),
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: there should be the option to just give it the settings

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+        }
+        return new SwerveModuleState(targetSpeed, Rotation2d.fromDegrees(targetAngle));
+    }
+
+    public double getSpeed() {
+        return driveEncoder.getVelocity();
+    }
+
+    /**
+     * Takes the module's angle and the desired angle, and returns it in within the range of 0 to 360.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: not what it does

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                },
+                        b -> turnController.stopMotor(), () -> false));
+        namespace.putData("move",
+                new FunctionalCommand(() -> set(new SwerveModuleState(1.5, Rotation2d.fromDegrees(0)), false), () -> {
+                },
+                        b -> turnController.stopMotor(), () -> false));
+        namespace.putRunnable("stop", turnController::stopMotor);
+        namespace.putNumber("applied output", driveController::getAppliedOutput);
+    }
+}
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: image

File: src/main/java/frc/robot/subsystems/SwerveModuleHolder.java [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+import com.revrobotics.CANSparkMax;
+import com.spikes2212.control.FeedForwardSettings;
+import com.spikes2212.control.PIDSettings;
+import com.spikes2212.dashboard.RootNamespace;
+import frc.robot.RobotMap;
+
+// https://www.youtube.com/watch?v=jxcMRkqaQdw
+public class SwerveModuleHolder {
+
+    private static final RootNamespace namespace = new RootNamespace("swerve module holder 2");
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModuleHolder.java]: 2?

File: src/main/java/frc/robot/subsystems/SwerveModuleHolder.java [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+    private static final boolean BACK_LEFT_DRIVE_INVERTED = false;
+    private static final boolean BACK_RIGHT_DRIVE_INVERTED = true;
+
+    private static final boolean FRONT_LEFT_CANCODER_INVERTED = true;
+    private static final boolean FRONT_RIGHT_CANCODER_INVERTED = true;
+    private static final boolean BACK_LEFT_CANCODER_INVERTED = true;
+    private static final boolean BACK_RIGHT_CANCODER_INVERTED = true;
+
+    private static final PIDSettings drivePIDSettings = namespace.addPIDNamespace("drive",
+            new PIDSettings(0, 0, 0, 0, 0));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModuleHolder.java]: change everything to -1 so it crashes

File: src/main/java/frc/robot/subsystems/SwerveModuleHolder.java [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+    private static final PIDSettings turnPIDSettings = namespace.addPIDNamespace("turn",
+            new PIDSettings(0.007, 0, 0, 0, 0));
+    private static final FeedForwardSettings driveFeedForwardSettings = namespace.addFeedForwardNamespace("drive",
+            new FeedForwardSettings(0, 0, 0, 0));
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModuleHolder.java]: ditto

File: vendordeps/Phoenix6.json [hunk 14]
```diff
@@ -336,4 +336,4 @@
       "simMode": "swsim"
     }
   ]
-}
+}
```
Comment (Yoel-Perman-Brilliant) [inline vendordeps/Phoenix6.json]: image

File: src/main/java/frc/robot/commands/DriveSwerve.java [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+    public DriveSwerve(Drivetrain drivetrain, Supplier<Double> xSpeed, Supplier<Double> ySpeed,
+                       Supplier<Double> rotationSpeed, boolean fieldRelative, boolean usePID) {
+        addRequirements(drivetrain);
+        this.drivetrain = drivetrain;
+        this.xSpeed = xSpeed;
+        this.ySpeed = ySpeed;
+        this.rotationSpeed = rotationSpeed;
+        this.fieldRelative = fieldRelative;
+        this.usePID = usePID;
+        this.xLimiter = new SlewRateLimiter(DRIVE_ACCELERATION_LIMIT_PERCENTAGE);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/DriveSwerve.java]: remove percentage

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    public void stop() {
+        frontLeft.stop();
+        frontRight.stop();
+        backLeft.stop();
+        backRight.stop();
+    }
+
+    public void pain() {
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: remove

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
+    public Rotation2d getRotation2d() {
+        return Rotation2d.fromDegrees(getAngle());
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putNumber("gyro yaw", this::getAngle);
+        namespace.putRunnable("stop drivetrain", () -> stop());
+        namespace.putRunnable("move PLEWA,SP;RKW.OIA", this::pain);
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: remove

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 18]
```diff
@@ -0,0 +1,227 @@
...
+        turnController.getPIDController().setReference(angle, CANSparkMax.ControlType.kPosition, PID_SLOT);
+    }
+
+    //speed - m/s
+    private void setSpeed(double speed, boolean usePID) {
+        if (usePID) {
+            configureDriveController();
+            configFF();
+            double feedForward = driveFeedForwardController.calculate(speed);
+            driveController.getPIDController().setReference(speed, CANSparkMax.ControlType.kVelocity, PID_SLOT,
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java:126]: make sure the feed forward is in voltages

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 19]
```diff
@@ -0,0 +1,227 @@
...
+                    feedForward);
+        } else driveController.set(speed / Drivetrain.MAX_SPEED_METERS_PER_SECONDS);
+    }
+
+    /**
+     * Minimize the change in heading the desired swerve module state would require by potentially
+     * reversing the direction the wheel spins. Customized from WPILib's version to include placing
+     * in appropriate scope for CTRE onboard control.
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: Rev not ctre

File: src/main/java/frc/robot/commands/DriveSwerve.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+import edu.wpi.first.math.filter.SlewRateLimiter;
+import edu.wpi.first.wpilibj2.command.Command;
+import frc.robot.subsystems.Drivetrain;
+
+import java.util.function.Supplier;
+
+public class DriveSwerve extends Command {
+
+    private static final double DRIVE_ACCELERATION_LIMIT = 2;
+    private static final double TURN_ACCELERATION_LIMIT_PERCENTAGE = 4;
```
Comment (Yoel-Perman-Brilliant) [inline src/main/java/frc/robot/commands/DriveSwerve.java]: no PRECENTAGE

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+    }
+
+    @Override
+    public void configureDashboard() {
+        namespace.putNumber("gyro yaw", this::getAngle);
+        namespace.putRunnable("stop drivetrain", () -> stop());
+        namespace.putNumber("x odom", () -> odometry.getPoseMeters().getX());
+        namespace.putNumber("y odom", () -> odometry.getPoseMeters().getY());
+        namespace.putNumber("rotation odom", () -> odometry.getPoseMeters().getRotation().getDegrees());
+
```
Comment (nettahalamish) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:132]: delete

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 22]
```diff
@@ -0,0 +1,228 @@
...
+            driveController.getPIDController().setReference(speed, CANSparkMax.ControlType.kVelocity, PID_SLOT,
+                    feedForward, SparkPIDController.ArbFFUnits.kVoltage);
+        } else driveController.set(speed / Drivetrain.MAX_SPEED_METERS_PER_SECONDS);
+    }
+
+    /**
+     * Minimize the change in heading the desired swerve module state would require by potentially
+     * reversing the direction the wheel spins. Customized from WPILib's version to include placing
+     * in appropriate scope for REV onboard control.
+     * Credit to team #364
```
Comment (nettahalamish) [inline src/main/java/frc/robot/subsystems/SwerveModule.java:151]: .

Comment (nettahalamish) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: .

File: src/main/java/frc/robot/subsystems/Drivetrain.java (hunk 21 above)
Comment (nettahalamish) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: delete

**Review by nettahalamish — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 23]
```diff
@@ -0,0 +1,228 @@
...
+    private void configureTurnController() {
+        turnController.setPeriodicFramePeriod(CANSparkLowLevel.PeriodicFrame.kStatus2, 40);
+        turnController.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        turnController.getPIDController().setP(turnPIDSettings.getkP());
+        turnController.getPIDController().setI(turnPIDSettings.getkI());
+        turnController.getPIDController().setD(turnPIDSettings.getkD());
+        turnController.setIdleMode(CANSparkMax.IdleMode.kCoast);
+    }
+
+    private void configureAbsoluteEncoder() {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: that's not a thing that's going to work

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Review by Spikes-2212 — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+    public static final Translation2d CENTER_OF_ROBOT = new Translation2d(
+            (FRONT_LEFT_WHEEL_POSITION.getX() + BACK_RIGHT_WHEEL_POSITION.getX()) / 2,
+            (FRONT_LEFT_WHEEL_POSITION.getY() + BACK_RIGHT_WHEEL_POSITION.getY()) / 2);
+
+    public static final double MAX_SPEED_METERS_PER_SECONDS = 4.7;
+    public static final double MIN_SPEED_METERS_PER_SECONDS = 0.2;
+
+    private static final String NAMESPACE_NAME = "drivetrain";
+
+    public final SwerveModule frontLeft;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: make all modules private

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 25]
```diff
@@ -0,0 +1,229 @@
...
+    private static final int PID_SLOT = 0;
+    private static final double STEERING_GEAR_RATIO = 12.8;
+    private static final double DRIVING_GEAR_RATIO = 6.12;
+    private static final double WHEEL_DIAMETER_IN_INCHES = 4;
+    private static final double INCHES_TO_METERS = 0.0254;
+    private static final double WHEEL_CIRCUMFERENCE_METERS = WHEEL_DIAMETER_IN_INCHES * INCHES_TO_METERS * Math.PI;
+    private static final double SECONDS_IN_MINUTE = 60;
+    private static final double DEGREES_IN_ROTATION = 360;
+
+    public final CANSparkMax driveController;
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: private both motors

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 26]
```diff
@@ -0,0 +1,229 @@
...
+        driveController.getPIDController().setD(drivePIDSettings.getkD());
+        driveController.setInverted(driveInverted);
+        driveController.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        driveEncoder.setVelocityConversionFactor(
+                ((1 / DRIVING_GEAR_RATIO) * WHEEL_CIRCUMFERENCE_METERS) / SECONDS_IN_MINUTE);
+        driveEncoder.setPositionConversionFactor((1 / DRIVING_GEAR_RATIO) * WHEEL_CIRCUMFERENCE_METERS);
+    }
+
+    private void configureTurnController() {
+        turnController.setPeriodicFramePeriod(CANSparkLowLevel.PeriodicFrame.kStatus2, 40);
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: image

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+        turnController.setIdleMode(CANSparkMax.IdleMode.kCoast);
+        turnController.getPIDController().setP(turnPIDSettings.getkP());
+        turnController.getPIDController().setI(turnPIDSettings.getkI());
+        turnController.getPIDController().setD(turnPIDSettings.getkD());
+        turnController.setIdleMode(CANSparkMax.IdleMode.kCoast);
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: bro why turn it coast twice

File: src/main/java/frc/robot/subsystems/SwerveModule.java [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+        }
+        if (newAngle - scopeReference > 180) {
+            newAngle -= 360;
+        } else if (newAngle - scopeReference < -180) {
+            newAngle += 360;
+        }
+        return newAngle;
+    }
+
+    public void configFF() {
```
Comment (Spikes-2212) [inline src/main/java/frc/robot/subsystems/SwerveModule.java]: put above private methods

**Review by Spikes-2212 — CHANGES_REQUESTED** [review]

**Review by Spikes-2212 — APPROVED** [review]

**Review by TurtleMinecraft — APPROVED** [review]

**Pull request merged**